iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/Catalogs.java

```diff
@@ -284,4 +334,42 @@ private static String getCatalogType(Configuration conf, String catalogName) {
       }
     }
   }
+
+  public static class TableCache {
+    private static final Cache<String, Table> tableCache = Caffeine.newBuilder()
+        .expireAfterAccess(12, TimeUnit.HOURS).build();
+    private static final Logger LOG = LoggerFactory.getLogger(TableCache.class);
+
+    public static void removeTables(Configuration conf) {
+      String queryId = conf.get(HiveConf.ConfVars.HIVEQUERYID.varname);
+      if (queryId != null && !queryId.isEmpty()) {
+        Set<String> queryKeys = tableCache.asMap().keySet().stream()
+            .filter(k -> k.startsWith(queryId)).collect(Collectors.toSet());
+        tableCache.invalidateAll(queryKeys);
+      } else {
+        LOG.warn("Query id is not present in config, therefore no Iceberg table object is removed " +
+            "from the table cache.");
+      }
+    }
+
+    public static Table getTable(Configuration conf, String catalogName, String tableIdentifier) {
+      String queryId = conf.get(HiveConf.ConfVars.HIVEQUERYID.varname);
+      if (queryId != null  && !queryId.isEmpty()) {
+        return tableCache.getIfPresent(getKey(queryId, catalogName, tableIdentifier));
+      }
+      return null;
+    }
+
+    public static void addTable(Configuration conf, String catalogName, String tableIdentifier, Table table) {
```

Comment:
Why not use lambda to generate the table object if not present? Like it is done with `putIfAbsent`?

Reply:
The `tableCache` doesn't have a put method that accepts functional interfaces.

Reply:
I was trying to suggest to add such a method to the `TableCache` class. I think it could make the code easier to read. What do you think?